lib/watir/window_collection.rb

```diff
@@ -0,0 +1,99 @@
+module Watir
+  class WindowCollection
+    include Enumerable
+    include Waitable
+
+    def initialize(browser, selector = {})
+      unless selector.keys.all? { |k| %i[title url element].include? k }
+        raise ArgumentError, "invalid window selector: #{selector.inspect}"
+      end
+
+      @browser = browser
+      @selector = selector
+    end
+
+    #
+    # Yields each window in collection.
+    #
+    # @yieldparam [Watir::Window]
+    #
+
+    def each(&blk)
+      reset!
+      to_a.each(&blk)
+    end
+
+    alias length count
+    alias size count
+    alias empty? none?
+
+    #
+    # First window of the collection
+    #
+    # @note windows in a collection are not ordered so this is not reliably
+    # @deprecated use Browser#switch_window or a better Window locator
+    # @return [Watir::Window] Returns an instance of a Watir::Window
+    #
+
+    def first
+      self[0]
+    end
+
+    #
+    # Last window of the collection
+    #
+    # @note windows in a collection are not ordered so this is not reliably
+    # @deprecated use Browser#switch_window or a better Window locator
+    # @return [Watir::Window] Returns an instance of a Watir::Window
+    #
+
+    def last
+      self[-1]
+    end
+
+    #
+    # Get the element at the given index or range.
+    #
+    # @note windows in a collection are not ordered so this is not reliably
+    # @deprecated use Browser#switch_window or a better Window locator
+    # @param [Integer, Range] value Index (0-based) or Range of desired element(s)
+    # @return [Watir::Element, Watir::ElementCollection] Returns an instance of a Watir::Element subclass
+    #
+
+    def [](value)
+      array = to_a
+
+      if array.size > 1
+        old = 'using indexing with windows'
+        new = 'Browser#switch_window if only 2 windows, ' \
+'or Browser#window with :title, :url or :element parameters if more than two windows'
+        reference = 'http://watir.com/window_indexes'
+        Watir.logger.deprecate old, new, reference: reference, ids: [:window_index]
+      end
+
+      to_a[value]
+    end
+
+    def ==(other)
+      to_a == other.to_a
+    end
+    alias eql? ==
+
+    def to_a
+      @to_a ||= begin
+        all = @browser.driver.window_handles.map { |wh| Window.new(@browser, handle: wh) }
+        if @selector.empty?
+          all
+        else
+          all.select do |win|
+            @selector.all? { |key, value| win.send(key) =~ /#{value}/ }
```

Comment:
As per the spec failure, this doesn't know how to handle the `:element` selector.

Reply:
Yeah, took me a couple tries and several more specs to make sure I had this right. :)